File: ios-native/NeobankNative/Sources/Core/Networking/APIClient.swift
import Foundation

enum HTTPMethod: String, Sendable {
    case get = "GET"
    case post = "POST"
    case put = "PUT"
    case patch = "PATCH"
    case delete = "DELETE"
}

struct EmptyResponse: Decodable, Sendable {}

/// Talks to the gateway BFF. Two instances are wired up in practice: one with
/// `auth == nil` for `/v1/auth/login` and `/v1/auth/register` (which must not
/// try to attach or refresh a bearer token), and one with `auth` set for
/// every other call.
final class APIClient: Sendable {
    struct AuthContext: Sendable {
        let tokenProvider: @Sendable () -> String?
        let refresh: @Sendable () async throws -> String?
        let onSessionExpired: @Sendable () async -> Void
    }

    private let baseURL: URL
    private let session: URLSession
    private let auth: AuthContext?
    private let refreshCoordinator = RefreshCoordinator()

    private static let mutatingMethods: Set<HTTPMethod> = [.post, .put, .patch, .delete]

    private static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    init(baseURL: URL, session: URLSession = .shared, auth: AuthContext? = nil) {
        self.baseURL = baseURL
        self.session = session
        self.auth = auth
    }

    @discardableResult
    func send<Response: Decodable>(
        _ path: String,
        method: HTTPMethod = .get,
        query: [String: String?]? = nil,
        body: [String: Any?]? = nil
    ) async throws -> Response {
        try await send(path, method: method, query: query, body: body, isRetryAfterRefresh: false)
    }

    private func send<Response: Decodable>(
        _ path: String,
        method: HTTPMethod,
        query: [String: String?]?,
        body: [String: Any?]?,
        isRetryAfterRefresh: Bool
    ) async throws -> Response {
        let request = try makeRequest(path: path, method: method, query: query, body: body)

        let data: Data
        let httpResponse: HTTPURLResponse
        do {
            let (responseData, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw APIError.network() }
            data = responseData
            httpResponse = http
        } catch let error as APIError {
            throw error
        } catch let error as URLError where error.code == .timedOut {
            throw APIError.timeout()
        } catch {
            throw APIError.network()
        }

        if httpResponse.statusCode == 401,
           let auth,
           !path.contains("/v1/auth/"),
           !isRetryAfterRefresh {
            guard (try? await refreshCoordinator.refresh(auth.refresh)) != nil else {
                await auth.onSessionExpired()
                throw APIError.unauthenticated()
            }
            return try await send(path, method: method, query: query, body: body, isRetryAfterRefresh: true)
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw mapFailure(statusCode: httpResponse.statusCode, data: data, response: httpResponse)
        }

        if Response.self == EmptyResponse.self {
            return EmptyResponse() as! Response // swiftlint:disable:this force_cast
        }

        do {
            return try Self.decoder.decode(Response.self, from: data)
        } catch {
            throw APIError.decoding()
        }
    }

    private func makeRequest(
        path: String,
        method: HTTPMethod,
        query: [String: String?]?,
        body: [String: Any?]?
    ) throws -> URLRequest {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if let query {
            let items = query.compactMapValues { $0 }.map { URLQueryItem(name: $0.key, value: $0.value) }
            if !items.isEmpty { components.queryItems = items }
        }
        var request = URLRequest(url: components.url!)
        request.httpMethod = method.rawValue
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(UUID().uuidString, forHTTPHeaderField: "X-Correlation-Id")

        if Self.mutatingMethods.contains(method) {
            request.setValue(UUID().uuidString, forHTTPHeaderField: "Idempotency-Key")
        }
        if let body {
            let compacted = body.compactMapValues { $0 }
            request.httpBody = try? JSONSerialization.data(withJSONObject: compacted)
        }
        if let auth, let token = auth.tokenProvider() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    private func mapFailure(statusCode: Int, data: Data, response: HTTPURLResponse) -> APIError {
        if statusCode == 401 { return .unauthenticated() }

        var message = "Something went wrong. Please try again."
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let apiMessage = json["error"] as? String {
            message = apiMessage
        }
        let correlationId = response.value(forHTTPHeaderField: "x-correlation-id")
        return APIError(message: message, statusCode: statusCode, correlationId: correlationId)
    }
}

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/core/networking/ApiClient.kt
package dev.horobets.neobanknative.core.networking

import java.io.IOException
import java.net.SocketTimeoutException
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNamingStrategy
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

enum class HttpMethod(val rawValue: String) {
    GET("GET"),
    POST("POST"),
    PUT("PUT"),
    PATCH("PATCH"),
    DELETE("DELETE"),
}

@Serializable
object EmptyResponse

/**
 * Talks to the gateway BFF. Two instances are wired up in practice: one with
 * `auth == null` for `/v1/auth/login` and `/v1/auth/register` (which must
 * not try to attach or refresh a bearer token), and one with `auth` set for
 * every other call.
 */
@OptIn(ExperimentalSerializationApi::class)
class ApiClient(
    private val baseUrl: String,
    private val httpClient: OkHttpClient = sharedHttpClient,
    private val auth: AuthContext? = null,
) {
    class AuthContext(
        val tokenProvider: () -> String?,
        val refresh: suspend () -> String?,
        val onSessionExpired: suspend () -> Unit,
    )

    private val refreshCoordinator = RefreshCoordinator()

    suspend inline fun <reified T> send(
        path: String,
        method: HttpMethod = HttpMethod.GET,
        query: Map<String, String?>? = null,
        body: JsonObject? = null,
        idempotencyKey: String? = null,
    ): T {
        val responseBody = executeRequest(path, method, query, body, idempotencyKey)
        if (T::class == EmptyResponse::class) {
            @Suppress("UNCHECKED_CAST")
            return EmptyResponse as T
        }
        return try {
            json.decodeFromString(responseBody)
        } catch (e: Exception) {
            throw ApiError.decoding()
        }
    }

    suspend fun executeRequest(
        path: String,
        method: HttpMethod,
        query: Map<String, String?>?,
        body: JsonObject?,
        idempotencyKey: String?,
    ): String = executeRequest(path, method, query, body, idempotencyKey, isRetryAfterRefresh = false)

    private suspend fun executeRequest(
        path: String,
        method: HttpMethod,
        query: Map<String, String?>?,
        body: JsonObject?,
        idempotencyKey: String?,
        isRetryAfterRefresh: Boolean,
    ): String = withContext(Dispatchers.IO) {
        val request = buildRequest(path, method, query, body, idempotencyKey)

        val response: Response = try {
            httpClient.newCall(request).execute()
        } catch (e: SocketTimeoutException) {
            throw ApiError.timeout()
        } catch (e: IOException) {
            throw ApiError.network()
        }

        response.use { resp ->
            val bodyString = resp.body?.string().orEmpty()

            if (resp.code == 401 && auth != null && !path.contains("/v1/auth/") && !isRetryAfterRefresh) {
                val refreshed = try {
                    refreshCoordinator.refresh(auth.refresh)
                } catch (e: Exception) {
                    null
                }
                if (refreshed == null) {
                    auth.onSessionExpired()
                    throw ApiError.unauthenticated()
                }
                return@withContext executeRequest(path, method, query, body, idempotencyKey, isRetryAfterRefresh = true)
            }

            if (resp.code !in 200..299) {
                throw mapFailure(resp.code, bodyString, resp)
            }

            bodyString
        }
    }

    private fun buildRequest(
        path: String,
        method: HttpMethod,
        query: Map<String, String?>?,
        body: JsonObject?,
        idempotencyKey: String?,
    ): Request {
        val urlBuilder = baseUrl.toHttpUrl().newBuilder()
        path.trim('/').split("/").forEach { segment ->
            if (segment.isNotEmpty()) urlBuilder.addPathSegment(segment)
        }
        query?.forEach { (key, value) -> if (value != null) urlBuilder.addQueryParameter(key, value) }

        val requestBuilder = Request.Builder().url(urlBuilder.build())
        requestBuilder.addHeader("Content-Type", "application/json")
        requestBuilder.addHeader("X-Correlation-Id", UUID.randomUUID().toString())
        if (method in mutatingMethods) {
            requestBuilder.addHeader("Idempotency-Key", idempotencyKey ?: UUID.randomUUID().toString())
        }
        auth?.tokenProvider?.invoke()?.let { token -> requestBuilder.addHeader("Authorization", "Bearer $token") }

        val requestBody = if (method == HttpMethod.GET) {
            null
        } else {
            val payload = body ?: JsonObject(emptyMap())
            json.encodeToString(JsonObject.serializer(), payload).toRequestBody(jsonMediaType)
        }
        requestBuilder.method(method.rawValue, requestBody)
        return requestBuilder.build()
    }

    private fun mapFailure(statusCode: Int, bodyString: String, response: Response): ApiError {
        if (statusCode == 401) return ApiError.unauthenticated()

        var message = "Something went wrong. Please try again."
        try {
            val root = json.parseToJsonElement(bodyString)
            val apiMessage = (root as? JsonObject)?.get("error") as? JsonPrimitive
            apiMessage?.contentOrNull?.let { message = it }
        } catch (e: Exception) {
            // Non-JSON or unexpected body shape — fall back to the generic message.
        }
        val correlationId = response.header("x-correlation-id")
        return ApiError(message = message, statusCode = statusCode, correlationId = correlationId, responseBody = bodyString)
    }

    companion object {
        private val mutatingMethods = setOf(HttpMethod.POST, HttpMethod.PUT, HttpMethod.PATCH, HttpMethod.DELETE)
        private val jsonMediaType = "application/json".toMediaType()

        /** convertFromSnakeCase equivalent — the gateway's field naming maps directly, no per-model CodingKeys/@SerialName. */
        val json: Json = Json {
            ignoreUnknownKeys = true
            namingStrategy = JsonNamingStrategy.SnakeCase
        }

        val sharedHttpClient: OkHttpClient by lazy {
            OkHttpClient.Builder()
                .connectTimeout(15, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build()
        }
    }
}
